File: content/snippets/cpp/language_basics/heaps_and_priority_queues.cpp
#include <iostream>
#include <queue>
#include <vector>

int main() {
    // max-heap (default)
    std::priority_queue<int> maxHeap;
    for (int v : {3, 1, 4, 1, 5, 9}) maxHeap.push(v);
    std::cout << "max: " << maxHeap.top() << "\n";  // 9

    // min-heap via greater<>
    std::priority_queue<int, std::vector<int>, std::greater<int>> minHeap;
    for (int v : {3, 1, 4, 1, 5, 9}) minHeap.push(v);
    std::cout << "drain min-heap: ";
    while (!minHeap.empty()) {
        std::cout << minHeap.top() << ' ';
        minHeap.pop();
    }
    std::cout << "\n";  // 1 1 3 4 5 9
    return 0;
}

File: content/snippets/scala3/language_basics/heaps_and_priority_queues.scala
import scala.collection.mutable

object HeapsAndPriorityQueues:
  @main def main(): Unit =
    val data = List(3, 1, 4, 1, 5, 9)

    // max priority queue (default)
    val maxPQ = mutable.PriorityQueue[Int]()
    data.foreach(maxPQ.enqueue)
    println(s"max: ${maxPQ.head}")  // 9

    // min priority queue via reverse ordering
    val minPQ = mutable.PriorityQueue[Int]()(Ordering[Int].reverse)
    data.foreach(minPQ.enqueue)
    print("drain min-heap: ")
    while (minPQ.nonEmpty) print(s"${minPQ.dequeue()} ")
    println()  // 1 1 3 4 5 9
